Enable nightly build with reproducible comparison

sophia-guo opened the pull request:
Related https://github.com/adoptium/temurin-build/issues/3245

Signed-off-by: Sophia Guo <sophia.gwf@gmail.com>

sxa commented:
Thanks for the explaination. Since this job isn't being auto-generated I would suggest that it may be preferable to keep it outside the normal jdk17u folder to ensure it doesn't get accidentally deleted at any point (someone might think that everythign in there can be regenerated automatically).

sxa commented on pipelines/build/common/openjdk_build_pipeline.groovy:
```diff
@@ -585,6 +585,30 @@ class Build {
+   def compareReproducibleBuild() {
+        //Currently only enable for jdk17+, linux_x64, temurin, nightly, which shouldn't affect current build
+        def jobName = "${env.JOB_NAME}_reproduce_compare"
+        if (getJavaVersionNumber() >= 17 &&
```
At the moment it looks like we only have the jobs for 17 - should this be changed to equals instead of >= ?

sophia-guo replied:
Yes, updated.

karianna commented on tools/reproduce_comparison/Jenkinsfile:
```diff
@@ -1,56 +1,134 @@
+                    def scmCommit = reproducedParams["OpenJDK Source Commit"]
+                    scmCommit = scmCommit.substring(scmCommit.lastIndexOf('/')+1)
+                    
+                    //convert COMPARED_JOB_PARAMS to json formatt string
```
```suggestion
                    // Convert COMPARED_JOB_PARAMS to json formatted string
```

karianna commented on pipelines/build/common/openjdk_build_pipeline.groovy:
```diff
@@ -1881,6 +1908,9 @@ class Build {
                 if (enableTests) {
                     try {
+                        // Compare reproducible build, using same build pipeline with enableTests = false to avoid recursive building
+                        compareReproducibleBuild()
```
Do we want this behind a enableReproducibleCheck flag?

sophia-guo replied:
During the initial step the reproducible comparison is limited to nightly jdk17, linux-x64, temurin build only https://github.com/adoptium/ci-jenkins-pipelines/pull/714/files#diff-b4bb07e2f448f9ad2986d69423ad0b704dcdc4d217d947f465e6ace3fde7a4b3R594. Going forward might need to a flag setting per jdk version and platform.

sophia-guo replied:
Updated. https://github.com/adoptium/ci-jenkins-pipelines/pull/714/commits/90635ee37a1afbd1737da81f036b6869b0f1ac4e

karianna commented on tools/reproduce_comparison/Jenkinsfile:
```diff
@@ -1,56 +1,134 @@
 /*
-Jenkins job does comparison for two different build artifacts
-If they are identical, build pass; or fail
+Jenkins job does reproduciable build compare.
```
```suggestion
Jenkins job does reproducible build compare.
```

karianna commented on tools/reproduce_comparison/Jenkinsfile:
```diff
@@ -1,56 +1,134 @@
+                        text(name: 'DEFAULTS_JSON', value: "${jsonJobParams.DEFAULTS_JSON}"),
+                        text(name: 'ADOPT_DEFAULTS_JSON', value: "${jsonJobParams.ADOPT_DEFAULTS_JSON}")
+                    ]
+                    def reproduciableJob = build job: "${COMPARED_JOB_NAME}",
```
```suggestion
                    def reproducibleJob = build job: "${COMPARED_JOB_NAME}",
```

karianna commented on tools/reproduce_comparison/Jenkinsfile:
```diff
@@ -1,56 +1,134 @@
+                            propagate: false,
+                            parameters: buildParams
+                    
+                    def result = reproduciableJob.getResult()
```
```suggestion
                    def result = reproducibleJob.getResult()
```

karianna commented on tools/reproduce_comparison/Jenkinsfile:
```diff
@@ -1,56 +1,134 @@
+                            timeout(time: 2, unit: 'HOURS') {
+                                copyArtifacts(
+                                    projectName:"${COMPARED_JOB_NAME}",
+                                    selector:specific("${reproduciableJob.getNumber()}"),
```
```suggestion
                                    selector:specific("${reproducibleJob.getNumber()}"),
```

karianna commented on tools/reproduce_comparison/Jenkinsfile:
```diff
@@ -1,56 +1,134 @@
-                        } else {
-                            echo 'Success: two builds are the same!'
+                    } else {
+                        error "Reproduced build failed, exit comparision"
```
hint to what the failed parts are?

sophia-guo replied:
Add the jenkins job name. There is job link in the job console output.

karianna commented on tools/reproduce_comparison/Jenkinsfile:
```diff
@@ -1,56 +1,134 @@
+                    def result2 = sh returnStatus: true, script: "tar -xz ${reproducedJDKFile[0].name} -C reproduced"
+                    if (result1 != 0 || result2 != 0) {
+                        currentBuild.result = 'UNSTABLE'
+                        error 'Wrong jdk tar files'
```
worth adding what those tar files are

karianna commented on tools/reproduce_comparison/JenkinsfileOld:
```diff
@@ -0,0 +1,81 @@
+/*
```
Do we just remove this file?

sxa commented:
@sophia-guo Can you avoid force pushing on every fix please since it makes it impossible to see what's changed since a reviewer last looked at the code.

sophia-guo commented:
Did a rebase and some changes are accidentally gone. 

> Can you avoid force pushing on every fix please since it makes it impossible to see what's changed since a reviewer last looked at the code.

@sxa I think with force push the changes since last review can still be available
https://github.com/adoptium/ci-jenkins-pipelines/compare/73a87bb1a396b637c8bab08c5a852ac54abaac29..2be0fecab230acc143eade49bb0eb500c935d1ee

andrew-m-leonard commented on tools/reproduce_comparison/Jenkinsfile:
```diff
@@ -1,56 +1,139 @@
-                                error 'Stopping after download and uncompress tar file'
+                    def sbomFiles = findFiles(glob: "**/*.json")
+                    def sbomParams = readJSON file: "${sbomFiles[0].name}"
+                    def buildTimeStamp = sbomParams.metadata.timestamp
```
Rather confusing I know, but sbomParams.metadata.timestamp is the CycloneDX sbom update "timestamp", not the Build timestamp, you need to get "Build Timestamp" in your loop below .. from the sbomParams.components[0].properties

andrew-m-leonard replied:
eg:
```
        {
          "name" : "Build Timestamp",
          "value" : "2023-06-07 03:31:17"
        },
```

sophia-guo replied:
> Rather confusing I know, but sbomParams.metadata.timestamp is the CycloneDX sbom update "timestamp", not the Build timestamp, you need to get "Build Timestamp" in your loop below .. from the sbomParams.components[0].properties

I was going to double check it with you about that. Thanks!

sophia-guo replied:
> Note, I suspect you'll need to craft that into ISO8601, it is UTC. Wondering if we should fix the SBOM so it is stored in ISO8601? https://github.com/adoptium/temurin-build/blob/7045470523b810ec64e580139a431b62f505e65d/sbin/build.sh#L791

Sorry, I'm not sure about this note. Current it is in ISO8601 format UTC time [buildTimestamp=$(date -u +"%Y-%m-%d %H:%M:%S") ](https://github.com/adoptium/temurin-build/blob/master/sbin/build.sh#L283). What is the correct time and correct format?

andrew-m-leonard replied:
yes agree

andrew-m-leonard commented on tools/reproduce_comparison/Jenkinsfile:
```diff
@@ -1,56 +1,139 @@
+                        dir('temurin-build/tools/reproduce_comparison') {
+                            def rc = 0
+                            if (COMPARED_JOB_NAME.contains('linux')) {
+                                sh "chmod 776 ./tooling/linux_repo_compare.sh"
```
Do we need to fix the linux_repo_compare.sh persmissions in github?

sxa replied:
776 (which makes it writable by anyone) seems a very odd number to choose here. Any reason why it shouldn't just be `chmod u+rx`? But if it can be resolved in GitHub that would be even better.


Or just skip the chmod and run `bash ./tooling/linux_repo_compare.sh` when you run it so the permissions don't really matte.

sophia-guo replied:
I didn't try this. Maybe it's an issue before as original jenkinfile has this fix https://github.com/adoptium/ci-jenkins-pipelines/blob/master/tools/reproduce_comparison/Jenkinsfile#L72

andrew-m-leonard reviewed (commented):
just a couple of comments

karianna commented on docs/UsingOurScripts.md:
```diff
@@ -78,6 +78,8 @@ This file contains the default constants and paths used in the build scripts for
     // Test suite and inclusion details
     "testDetails"            : {
+        // Boolean denoting wether the reproducible compare is needed
+        "enableReproducibleCompare" :false,
```
```suggestion
        "enableReproducibleCompare" : false,
```

karianna commented on pipelines/build/common/build_base_file.groovy:
```diff
@@ -213,7 +215,22 @@ class Builder implements Serializable {
+        }
+        return enableReproducibleCompare
+    }
     /*
```
```suggestion

    /*
```

karianna commented on pipelines/build/common/config_regeneration.groovy:
```diff
@@ -334,7 +334,24 @@ class Regeneration implements Serializable {
+        }
+        return enableReproducibleCompare
+    }
     /*
```
```suggestion

    /*
```

karianna commented on pipelines/jobs/release_pipeline_job_template.groovy:
```diff
@@ -77,6 +78,7 @@ pipelineJob("${BUILD_FOLDER}/${JOB_NAME}") {
         stringParam('baseFilePath', '', "Relative path to where the build_base_file.groovy file is located. This runs the downstream job setup and configuration retrieval services.<br>Default: <code>${defaultsJson['baseFileDirectories']['upstream']}</code>")
         stringParam('buildConfigFilePath', '', "Relative path to where the jdkxx_pipeline_config.groovy file is located. It contains the build configurations for each platform, architecture and variant.<br>Default: <code>${defaultsJson['configDirectories']['build']}/jdkxx_pipeline_config.groovy</code>")
         booleanParam('aqaAutoGen', false, 'If set to true, force auto generate AQA test jobs. Defaults to false')
+        booleanParam('enableReproducibleCompare', runReproducibleCompare, 'If set to true the reproducible compare job might be triggerred')
```
```suggestion
        booleanParam('enableReproducibleCompare', runReproducibleCompare, 'If set to true the reproducible compare job might be triggered')
```

karianna commented on tools/reproduce_comparison/Jenkinsfile:
```diff
@@ -1,56 +1,139 @@
+
+                    def originalJDKFile = findFiles(glob: "original/*.tar.gz")
+                    def reproducedJDKFile = findFiles(glob: "reproduced/*.tar.gz")
+                    def result1 = sh returnStatus: true, script: "tar -xz ${originalJDKFile[0].name} -C original"
```
```suggestion
                    def originalBuildResult = sh returnStatus: true, script: "tar -xz ${originalJDKFile[0].name} -C original"
```

karianna commented on tools/reproduce_comparison/Jenkinsfile:
```diff
@@ -1,56 +1,139 @@
+                    def originalJDKFile = findFiles(glob: "original/*.tar.gz")
+                    def reproducedJDKFile = findFiles(glob: "reproduced/*.tar.gz")
+                    def result1 = sh returnStatus: true, script: "tar -xz ${originalJDKFile[0].name} -C original"
+                    if (result1 != 0 ) {
```
```suggestion
                    if (originalBuildResult != 0 ) {
```

karianna commented on tools/reproduce_comparison/Jenkinsfile:
```diff
@@ -1,56 +1,139 @@
+                        error " Untar ${originalJDKFile[0].name} failed"
+                        return
+                    } 
+                    def result2 = sh returnStatus: true, script: "tar -xz ${reproducedJDKFile[0].name} -C reproduced"
```
```suggestion
                    def reproducedBuildResult = sh returnStatus: true, script: "tar -xz ${reproducedJDKFile[0].name} -C reproduced"
```

karianna commented on tools/reproduce_comparison/Jenkinsfile:
```diff
@@ -1,56 +1,139 @@
+                        return
+                    } 
+                    def result2 = sh returnStatus: true, script: "tar -xz ${reproducedJDKFile[0].name} -C reproduced"
+                    if (result2 != 0 ) {
```
```suggestion
                    if (reproducedBuildResult != 0 ) {
```

karianna commented:
Some more linter and test fixes needed

sxa commented on docs/UsingOurScripts.md:
```diff
@@ -79,7 +79,7 @@ This file contains the default constants and paths used in the build scripts for
     // Test suite and inclusion details
     "testDetails"            : {
         // Boolean denoting wether the reproducible compare is needed
```
```suggestion
        // Boolean denoting whether the reproducible compare is needed
```

sophia-guo commented:
Note the test failed at https://github.com/adoptium/ci-jenkins-pipelines/pull/714/files#diff-fc15f6bbd3ef3cf595558447270454056407397071c7efbd13ca516fcf0a3a5bR66. Test is using master default.json, which has not been updated with this PR

sophia-guo commented:
Depends https://github.com/adoptium/temurin-build/pull/3386

andrew-m-leonard reviewed (approved):
nice work @sophia-guo

andrew-m-leonard reviewed (changes requested):
Just an update to https://raw.githubusercontent.com/adoptium/ci-jenkins-pipelines/master/pipelines/defaults.json required

andrew-m-leonard commented on pipelines/src/test/groovy/RepoHandlerTest.groovy:
```diff
@@ -63,6 +63,7 @@ class RepoHandlerTest {
         Assertions.assertEquals(adoptJson.defaultsUrl, 'https://raw.githubusercontent.com/adoptium/ci-jenkins-pipelines/master/pipelines/defaults.json')
 
         // Test details
+        Assertions.assertTrue(adoptJson.testDetails.enableReproducibleCompare instanceof Boolean)
```
ah you have, I can't read !

andrew-m-leonard replied:
CI "Test" fails as RepoHandler points at master... so just ignore that failure.

andrew-m-leonard reviewed (approved):
lgtm

andrew-m-leonard commented:
> Just an update to https://raw.githubusercontent.com/adoptium/ci-jenkins-pipelines/master/pipelines/defaults.json required

Has been updated...

sophia-guo commented:
> https://github.com/adoptium/ci-jenkins-pipelines/pull/714#issuecomment-1587651453

Shouldn't. See 
https://github.com/adoptium/ci-jenkins-pipelines/pull/714#issuecomment-1583101305
